engines/dfc_provider/app/services/product_type_importer.rb

```diff
@@ -1,11 +1,31 @@
 # frozen_string_literal: true
 
 class ProductTypeImporter < DfcBuilder
+  # Try to find the taxon closest matching the given product type.
+  # If we don't find any matching taxon, we return a random one.
   def self.taxon(product_type)
-    dfc_id = product_type&.semanticId
+    priority_list = [product_type, *list_broaders(product_type)].compact
 
-    # Every product needs a primary taxon to be valid. So if we don't have
-    # one or can't find it we just take a random one.
-    Spree::Taxon.find_by(dfc_id:) || Spree::Taxon.first
+    # Optimistic querying.
+    # We could query all broader taxons in one but then we need to still sort
+    # them locally and use more memory. That would be a pessimistic query.
+    # Consider caching the result instead.
+    taxons = priority_list.lazy.map do |type|
+      Spree::Taxon.find_by(dfc_id: type.semanticId)
+    end.compact
+
+    taxons.first || Spree::Taxon.first
```

Comment:
Nice to see different ways of doing things. I probably would have ended up with an `each` or `reduce` loop, but this is nice 👍